ambry-store/src/main/java/com.github.ambry.store/BlobStoreCompactor.java

```diff
@@ -733,53 +761,93 @@ private boolean copyRecords(LogSegment logSegmentToCopy, List<IndexEntry> srcInd
                       srcValue.getOperationTimeInMs(), srcValue.getAccountId(), srcValue.getContainerId());
-              tgtValue.setFlag(IndexValue.Flags.Delete_Index);
-              tgtValue.clearOriginalMessageOffset();
               tgtIndex.addToIndex(new IndexEntry(srcIndexEntry.getKey(), tgtValue), fileSpan);
             }
+            long lastModifiedTimeSecsToSet =
+                srcValue.getOperationTimeInMs() != Utils.Infinite_Time ? srcValue.getOperationTimeInMs() / Time.MsPerSec
+                    : lastModifiedTimeSecs;
+            tgtIndex.getIndexSegments().lastEntry().getValue().setLastModifiedTimeSecs(lastModifiedTimeSecsToSet);
+            writtenLastTime = srcValue.getSize();
+            srcMetrics.compactionCopyRateInBytes.mark(srcValue.getSize());
+          } else if (!isActive) {
+            logger.info("Stopping copying in {} because shutdown is in progress", storeId);
+            copiedAll = false;
+            break;
           } else {
-            IndexValue tgtValue =
-                new IndexValue(srcValue.getSize(), fileSpan.getStartOffset(), srcValue.getExpiresAtMs(),
-                    srcValue.getOperationTimeInMs(), srcValue.getAccountId(), srcValue.getContainerId());
-            tgtIndex.addToIndex(new IndexEntry(srcIndexEntry.getKey(), tgtValue), fileSpan);
+            // this is the extra segment, so it is ok to run out of space.
+            logger.info(
+                "There is no more capacity in the destination log in {}. Total capacity is {}. Used capacity is {}."
+                    + " Segment that was being copied is {}", storeId, totalCapacity, usedCapacity,
+                logSegmentToCopy.getName());
+            copiedAll = false;
+            break;
           }
-          long lastModifiedTimeSecsToSet =
-              srcValue.getOperationTimeInMs() != Utils.Infinite_Time ? srcValue.getOperationTimeInMs() / Time.MsPerSec
-                  : lastModifiedTimeSecs;
-          tgtIndex.getIndexSegments().lastEntry().getValue().setLastModifiedTimeSecs(lastModifiedTimeSecsToSet);
-          writtenLastTime = srcValue.getSize();
-          srcMetrics.compactionCopyRateInBytes.mark(srcValue.getSize());
-        } else if (!isActive) {
-          logger.info("Stopping copying in {} because shutdown is in progress", storeId);
-          copiedAll = false;
-          break;
-        } else {
-          // this is the extra segment, so it is ok to run out of space.
-          logger.info(
-              "There is no more capacity in the destination log in {}. Total capacity is {}. Used capacity is {}."
-                  + " Segment that was being copied is {}", storeId, totalCapacity, usedCapacity,
-              logSegmentToCopy.getName());
-          copiedAll = false;
+        }
+        if (!copiedAll) {
+          // break outer while loop
           break;
         }
+        startIndex = endIndex + 1;
```

Comment:
could you add `tgtLog.flush()` just before this line? It might help keep the dirty pages in the buffer cache to a minimum.